### PR TITLE
Sunny/choose deploy parameters

### PR DESCRIPTION
**Pull request by sunnycarter**
Help users choose which VNF or CNF parameters they want to expose as deploymentParameters on the NF.

Adds --order_params option for VNFs
Adds the ability to create the values mapping file for CNFs
Adds --interactive option for VNFs and CNFs to allow the user to select y/n for each parameter.
Tested build with fed-elastic CNF and SIMPLVM VNF but have not tested the results, nor have I tested multiple helm charts.

Also has some linting changes which are nothing to do with these changes, sorry reviewer.

### CODE REVIEW

File: src/aosm/azext_aosm/generate_nfd/vnf_nfd_generator.py [hunk 1]
```diff
@@ -147,19 +221,44 @@ def write_deployment_parameters(self, folder_path: str) -> None:
...
+                )
+                with open(
+                    optional_deployment_parameters_path, "w", encoding="utf-8"
+                ) as _file:
+                    _file.write(OPTIONAL_DEPLOYMENT_PARAMETERS_HEADING)
+                    _file.write(json.dumps(nfd_parameters_with_default, indent=4))
+                print(
+                    "Optional ARM parameters detected. Created "
+                    f"{OPTIONAL_DEPLOYMENT_PARAMETERS_FILE} to help you choose which "
+                    "to expose."
```
[the remark this answers is not in the document]
Comment (jamiedparsons) [reply src/aosm/azext_aosm/generate_nfd/vnf_nfd_generator.py:244]: Ah the heading in the file explains

**Review by jamiedparsons — COMMENTED** [review]
Thanks for doing this Sunny - they are great new features.  I've left a bunch of comments, but a lot of that was just me trying to get my head around what you were doing.

Feel free to grab me if anything isn't clear.

File: src/aosm/azext_aosm/_configuration.py [hunk 2]
```diff
@@ -60,7 +62,12 @@
     "path_to_chart":
         "File path of Helm Chart on local disk. Accepts .tgz, .tar or .tar.gz",
     "path_to_mappings":
-        "File path of value mappings on local disk. Accepts .yaml or .yml",
+        "File path of value mappings on local disk where chosen values are replaced "
+        "with deploymentParameter placeholders. Accepts .yaml or .yml. If left as a "
+        "blank string, a value mappings file will be generated with every value "
+        "mapped to a deployment parameter. Use a blank string and --interactive on "
+        "the build command to interactively choose which values to map."
```
[the remark this answers is not in the document]
Comment (sunnycarter) [reply src/aosm/azext_aosm/_configuration.py:69]: Let's have a quick discussion

File: src/aosm/azext_aosm/custom.py [hunk 3]
```diff
@@ -90,13 +99,17 @@ def _get_config_from_file(
     return config
 
 
-def _generate_nfd(definition_type, config):
+def _generate_nfd(
+    definition_type: str, config: NFConfiguration, order_params: bool, interactive: bool
+):
     """Generate a Network Function Definition for the given type and config."""
     nfd_generator: NFDGenerator
     if definition_type == VNF:
```
[the remark this answers is not in the document]
Comment (sunnycarter) [reply src/aosm/azext_aosm/custom.py:107]: no, can't be bothered :)

File: src/aosm/azext_aosm/generate_nfd/cnf_nfd_generator.py [hunk 4]
```diff
@@ -390,17 +524,27 @@ def find_deploy_params(
...
-                final_schema.update(
-                    {param: {"type": schema_nested_dict["properties"][k]["type"]}}
-                )
-
+                if "properties" in schema_nested_dict.keys():
+                    final_schema.update(
+                        {param: {"type": schema_nested_dict["properties"][k]["type"]}}
+                    )
+                else:
+                    final_schema.update(
```
[the remark this answers is not in the document]
Comment (sunnycarter) [reply src/aosm/azext_aosm/generate_nfd/cnf_nfd_generator.py:542]: I'm a bit confused by this whole section. Jordan got a lot of code review comments from chaos on it but she couldn't understand them so hasn't implemented them yet. I just fixed a bug I found, but have added some comments in an attempt to decipher what is going on.  It seems to work but I think is a bit flawed.

File: src/aosm/azext_aosm/generate_nfd/vnf_nfd_generator.py [hunk 5]
```diff
@@ -126,16 +171,45 @@ def write_deployment_parameters(self, folder_path: str) -> None:
...
-            elif self.vm_parameters[key]["type"] == "secureString":
-                nfd_parameters[key] = {"type": "string"}
-            else:
-                nfd_parameters[key] = {"type": self.vm_parameters[key]["type"]}
+        vm_parameters = (
+            self.vm_parameters_ordered if self.order_params else self.vm_parameters
+        )
+
+        for key in vm_parameters:
+            # Order parameters into those with and without defaults
```
[the remark this answers is not in the document]
Comment (sunnycarter) [reply src/aosm/azext_aosm/generate_nfd/vnf_nfd_generator.py:182]: no

File: src/aosm/azext_aosm/generate_nfd/vnf_nfd_generator.py [hunk 6]
```diff
@@ -126,16 +171,45 @@ def write_deployment_parameters(self, folder_path: str) -> None:
...
+                nfd_parameters_with_default[key] = {"type": json_type}
+
+            nfd_parameters[key] = {"type": json_type}
+
+        # Now we are out of the vm_parameters loop, we can remove the excluded
+        # parameters so they don't get included in templateParameters.json
+        # Remove from both ordered and unordered dicts
+        for key in vm_parameters_to_exclude:
+            self.vm_parameters.pop(key, None)
+            self.vm_parameters_ordered.pop(key, None)
```
[the remark this answers is not in the document]
Comment (sunnycarter) [reply src/aosm/azext_aosm/generate_nfd/vnf_nfd_generator.py:212]: have done so. It did work before but I agree was odd. Will test again.

**Pull request merged**